packages/adblocker-extended-selectors/src/eval.ts

```diff
@@ -71,104 +71,247 @@ export function matches(element: Element, selector: AST): boolean {
+  for (let i = 0; i < distance; i++) {
+    ancestor = ancestor.parentElement;
+    if (ancestor === null) {
+      return null;
+    }
+  }
+  return ancestor;
+}
 
-  if (
-    selector.type === 'id' ||
-    selector.type === 'class' ||
-    selector.type === 'type' ||
-    selector.type === 'attribute'
-  ) {
-    elements.push(...element.querySelectorAll(selector.content));
-  } else if (selector.type === 'list') {
-    for (const subSelector of selector.list) {
-      elements.push(...querySelectorAll(element, subSelector));
+function findAncestorBySelector(element: Element, selector: string): Element | null {
+  let ancestor: Element | null = element.parentElement;
+  while (ancestor !== null) {
+    if (ancestor.matches && ancestor.matches(selector)) {
+      return ancestor;
     }
-  } else if (selector.type === 'compound') {
-    // TODO - handling compound needs to be reworked...
-    // .cls:upward(1) for example will not work with this implementation.
-    // :upward is not about selecting, but transforming a set of nodes (i.e.
-    // uBO's transpose method).
-    if (selector.compound.length !== 0) {
-      elements.push(
-        ...querySelectorAll(element, selector.compound[0]).filter((e) =>
-          selector.compound.slice(1).every((s) => matches(e, s)),
-        ),
-      );
+    ancestor = ancestor.parentElement;
+  }
+  return null;
+}
+
+function handleCompoundSelector(element: Element, compound: AST[]): Element[] {
+  if (compound.length === 0) {
+    return [];
+  }
+
+  const upwardIndex = compound.findIndex((s) => s.type === 'pseudo-class' && s.name === 'upward');
+
+  if (upwardIndex === -1) {
+    const [firstSelector, ...restSelectors] = compound;
+    return querySelectorAll(element, firstSelector).filter((e) =>
+      restSelectors.every((s) => matches(e, s)),
+    );
+  }
+
+  const before = compound.slice(0, upwardIndex);
+  const upward = compound[upwardIndex];
+  const after = compound.slice(upwardIndex + 1);
+
+  const candidates =
+    before.length > 0
+      ? querySelectorAll(element, { type: 'compound', compound: before })
+      : [element];
+
+  const ancestors: Element[] = [];
+
+  for (const c of candidates) {
+    if (upward.type !== 'pseudo-class' || upward.name !== 'upward') {
+      continue;
     }
-  } else if (selector.type === 'complex') {
-    const elements2 =
-      selector.left === undefined ? [element] : querySelectorAll(element, selector.left);
-
-    if (selector.combinator === ' ') {
-      for (const element2 of elements2) {
-        elements.push(...querySelectorAll(element2, selector.right));
-      }
-    } else if (selector.combinator === '>') {
-      for (const element2 of elements2) {
-        for (const child of element2.children) {
-          if (matches(child, selector.right) === true) {
-            elements.push(child);
-          }
-        }
+    const { argument } = upward;
+    if (argument === undefined) {
+      continue;
+    }
+
+    const n = Number(argument);
```

Comment:
Nit. Do we want to make sure that `argument` is an integer here?

Reply:
I think `parseInt(argument, 10)` should be here instead of `Number`

Reply:
Alternatively, we can try `^\d+$` then parse number later